### DISCUSSION TITLE
PVE Tools > GPU Acceleration & Hardware Passthrough Setup Script

### DISCUSSION THREAD

**Discussion by ndoggac** [community-scripts/ProxmoxVE#759] [Ideas]
An additional script under PVE Tools to setup GPU & Hardware passthrough.

1.  Verify IOMMU is enabled in BIOS

2. Start with GPU(s).
 a. Identify all graphical chipsets from nvidia, amd, or intel. (grep VGA)
 b. Have user identify the primary display to ignore, or choose to disable it for the hypervisor to dedicate its resources to other (dangerous)
 c. Have user select IOMMU passthrough or SR-IOV distribution of GPU resources using hypervisor driver.
 d. Make changes to Proxmox to enable either IOMMU or SR-IOV
 e. Have user select which VM(s) or LXC(s) they want to pass-through a device to, or assign GPU resources to.
 f. If SR-IOV, install the appropriate drivers for the available chipset(s).

3.  Identify other PCIE or USB devices
 a. Have user select pass-through options, and assign them to appropriate VM or LXC.

**Comment by alexanderjerome** [comment]
This script handles some of that, apparently though I can't verify how well it works, but it's worked for me. It's the only one I've found, thus far. But I agree, proper passthrough automation would be tight. One challenge I see though is handling the process post-reboot, as you'll need to reboot at least once and then pick up afterwards. You would need some verification of which devices already have passthrough configured properly configured and means to verify.

https://github.com/Danilop95/Proxmox-Enhanced-Configuration-Utility

When it comes to GPU passthrough, you would need to automate installation of drivers. I've only done NVIDIA installation, and I do recall seeing a tool somewhere that can assist with that. Not sure about AMD, but I hear it's a bit more straightforward due to the tight Linux integration.

There's also the matter of device mapping, a relatively recent feature of Proxmox. I'm not sure if resource mapping can be configured via the .cfg files directly, but it would be an ideal way of letting the user know which devices were mapped with the utility.

You would also need to be careful not to overwrite any current kernel configurations in the grub. Some kind of failsafe / backup would be in order.

I've had some weird issues with blacklisting and vfio binding. Last time I installed the GPU drivers, I couldn't properly configure vfio unless I specifically binded the device ids in the grub. This was because I was configuring GPU passthrough so that I could use a VM with display output and USB controller passthrough. Worked like a charm, but took a while to configure.

My 2 cents, but yes I would love to see this.

**Comment by remz1337** [comment]
there's still a part that needs to be done manually (enabling IOMMU in the BIOS), but yes, further configuration could be automated. I also have a script in my fork that detects and automate Nvidia GPU passthrough, given the host is already configured:
https://github.com/remz1337/ProxmoxVE/blob/remz/misc/nvidia.func

It's mainly used with my frigate script. A few people have reported successful installation. Not sure if it will ever make it into the official repo, might be something to discuss with the other contributors

**Reply by MickLesk** [reply]
create an PR and we can discuss all :-)
